daemon/containerd/mount.go

```diff
@@ -0,0 +1,52 @@
+package containerd
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	"github.com/containerd/containerd/mount"
+	"github.com/docker/docker/container"
+	"github.com/docker/docker/pkg/containerfs"
+	"github.com/sirupsen/logrus"
+)
+
+// Mount mounts and sets container base filesystem
+func (i *ImageService) Mount(ctx context.Context, container *container.Container) error {
+	snapshotter := i.client.SnapshotService(container.Driver)
+	mounts, err := snapshotter.Mounts(ctx, container.ID)
+	if err != nil {
+		return err
+	}
+
+	tempMountLocation := os.TempDir()
+
+	root, err := os.MkdirTemp(tempMountLocation, "rootfs-mount")
+	if err != nil {
+		return fmt.Errorf("failed to create temp dir: %w", err)
+	}
+
+	if err := mount.All(mounts, root); err != nil {
+		return fmt.Errorf("failed to mount %s: %w", root, err)
+	}
+
+	container.BaseFS = containerfs.NewLocalContainerFS(root)
+	return nil
+}
+
+// Unmount unmounts the container base filesystem
+func (i *ImageService) Unmount(_ context.Context, container *container.Container) error {
+	root := container.BaseFS.Path()
+
+	if err := mount.UnmountAll(root, 0); err != nil {
+		return fmt.Errorf("failed to unmount %s: %w", root, err)
+	}
+
+	if err := os.Remove(root); err != nil {
+		logrus.WithError(err).WithField("dir", root).Error("failed to remove mount temp dir")
+	}
+
+	container.BaseFS = nil
```

Comment:
:warning: I noticed that this was not in the old code

perhaps that was a bug (should it be included as separate commit before "refactor"?) /cc @vvoland @rumpl

Reply:
This was intended. Since we already unmounted and removed the dir the BaseFS set here no longer makes sense.
It will also make it more visible if there is some code path that tries to use the BaseFS after unmount, because it will panic due to nil pointer instead of trying to perform some funny stuff on no longer existing path.

Reply:
Oh! sorry, I think I commented on the wrong file. I meant to comment on the `daemon/images/mount.go` (so, the graph driver implementation, not the snapshotter)

Reply:
Ah, yeah I think it would make sense to consider lack of resetting this to nil as a bug. In this case the directory may still exist, because the RWLayer.Unmount seems to just decrement the refcount for this mount, so something else may still use this.
But since Unmount was called on this container, I think it shouldn't have an access to this dir any longer.
+1 on extracting it to another PR, someone else from the maintainers may have some input on this so it would be nice not to sneak this change silently with this PR.

Reply:
Yes, I was wondering if it was used elsewhere (e.g., cleaning up dead containers that failed to be removed)

Reply:
This one is interesting actually.

The `daemon/images/mount.go` `Unmount` implementation shouldn't set the `BaseFS` to nil. The difference is that the graph driver's implementation mounts the rootfs to its final destination, but the snapshotter one creates a new mount each time. 
So even if you `Unmount` there is other code that uses the `BaseFS` of the container. For example when you `docker exec` we don't mount the rootfs but you look at the container's `BaseFS` and then look for resources inside (like the UID/GID for example). So if we set (in the graph driver's case) the `BaseFS` to nil on unmount parts of the code don't work any more.